Automattic/wp-calypso
# Checkout: use responseCart for PayPal transaction cart

#### Changes proposed in this Pull Request

The PayPal transaction endpoint requires a `cart` object with a specific schema. Currently the way that cart object is generated is that the shopping-cart endpoint data is converted into composite-checkout format and passed into `CheckoutProvider`; it is then extracted from the provider by the PayPal payment method button and passed to the payment processor function; from there it's parsed and converted to the transaction endpoint schema. This is inefficient, however, because the transaction endpoint schema is nearly the same as the shopping-cart endpoint data. This PR passes the shopping-cart data directly to the PayPal processor function instead. Once this is done for all payment processor functions, a lot of code can be removed.

To do
- [x] See if `google_apps_registration_data` is already part of the shopping-cart data, and if so, we don't need to pass it through `addRegistrationDataToGSuiteItem`.
- [x] Make a simpler version of `createTransactionEndpointCartFromLineItems` for use with a `ResponseCart` rather than overloading the current implementation. This is safer on several levels and more explicit.
- [ ] Remove passing `items` to the PayPal `onClick` processor function call.
- [x] Move `WPCOMTransactionEndpointCartItem` type fix to its own PR (done in https://github.com/Automattic/wp-calypso/pull/51126)
- [x] Verify that the transaction endpoint really needs `currency` set for each product, and if not, can we just use `RequestCartProduct` instead of `WPCOMTransactionEndpointCartItem`?

Requries https://github.com/Automattic/wp-calypso/pull/51084 to be merged first.

#### Testing instructions

- Make a purchase with the PayPal payment processor and verify that it works.
- Make a purchase of a G Suite product and verify that it works correctly (notably that the `google_apps_registration_data` is sent to the endpoint.
